// @kameleoon/javascript-sdk-core/dist/storage/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ClientDataType", {
  enumerable: true,
  get: function () {
    return _types.ClientDataType;
  }
});
Object.defineProperty(exports, "ExperimentVariationsType", {
  enumerable: true,
  get: function () {
    return _types.ExperimentVariationsType;
  }
});
Object.defineProperty(exports, "ExternalStorage", {
  enumerable: true,
  get: function () {
    return _externalStorage.ExternalStorage;
  }
});
Object.defineProperty(exports, "IExternalStorage", {
  enumerable: true,
  get: function () {
    return _types.IExternalStorage;
  }
});
Object.defineProperty(exports, "IExternalStorageConstructor", {
  enumerable: true,
  get: function () {
    return _types.IExternalStorageConstructor;
  }
});
Object.defineProperty(exports, "IStorage", {
  enumerable: true,
  get: function () {
    return _types.IStorage;
  }
});
Object.defineProperty(exports, "KameleoonStorageKey", {
  enumerable: true,
  get: function () {
    return _constants.KameleoonStorageKey;
  }
});
Object.defineProperty(exports, "VariationDataType", {
  enumerable: true,
  get: function () {
    return _types.VariationDataType;
  }
});
var _constants = require("./constants");
var _externalStorage = require("./externalStorage");
var _types = require("./types");
