// @kameleoon/javascript-sdk/dist/kameleoonClient.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.KameleoonClient = void 0;
var _javascriptSdkCore = require("@kameleoon/javascript-sdk-core");
var _storage = require("./storage");
var _requester = require("./requester");
var _eventSource = require("./eventSource");
/**
 * @class
 * KameleoonClient - a class for creating kameleoon client instance
 * */
class KameleoonClient extends _javascriptSdkCore.KameleoonCore {
  constructor(siteCode, configuration) {
    const storage = new _storage.StorageConstructor();
    const eventSource = new _eventSource.EventSourceConstructor();
    const requestDispatcher = new _requester.RequestDispatcher();
    super({
      siteCode,
      configuration,
      internalConfiguration: {
        externalStorage: storage,
        externalEventSource: eventSource,
        externalRequestDispatcher: requestDispatcher,
        externalPackageInfo: {
          type: _javascriptSdkCore.SDKType.JavaScript,
          version: "1.3.1"
        }
      }
    });
  }
}
exports.KameleoonClient = KameleoonClient;
