// @kameleoon/javascript-sdk-core/dist/kameleoonError/types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.KameleoonException = void 0;
let KameleoonException = /*#__PURE__*/function (KameleoonException) {
  KameleoonException["ExperimentConfigurationNotFound"] = "ExperimentConfigurationNotFound";
  KameleoonException["EventSourceInitialization"] = "EventSourceInitialization";
  KameleoonException["VariationNotFound"] = "VariationNotFound";
  KameleoonException["FeatureFlagConfigurationNotFound"] = "FeatureFlagConfigurationNotFound";
  KameleoonException["NotTargeted"] = "NotTargeted";
  KameleoonException["NotAllocated"] = "NotAllocated";
  KameleoonException["VisitorCodeMaxLength"] = "VisitorCodeMaxLength";
  KameleoonException["StorageInitialization"] = "StorageInitialization";
  KameleoonException["StorageWrite"] = "StorageWrite";
  KameleoonException["StorageRead"] = "StorageRead";
  KameleoonException["StorageParse"] = "StorageParse";
  KameleoonException["ClientConfiguration"] = "ClientConfiguration";
  KameleoonException["TargetingCondition"] = "TargetingCondition";
  KameleoonException["AmongValuesCheck"] = "AmongValuesCheck";
  KameleoonException["Initialization"] = "Initialization";
  KameleoonException["JSONParse"] = "JSONParse";
  KameleoonException["NumberParse"] = "NumberParse";
  KameleoonException["RemoteData"] = "RemoteData";
  return KameleoonException;
}({});
exports.KameleoonException = KameleoonException;
