// @kameleoon/javascript-sdk-core/dist/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.VISITOR_CODE_LENGTH = exports.UNALLOCATED_VARIATION_ID = exports.SDKType = exports.OFF_VARIATION_KEY = exports.KAMELEOON_COOKIE_KEY = exports.Cookie = exports.CACHE_ITEM_LIFETIME = exports.CACHE_CLEANUP_TIMEOUT = void 0;
const UNALLOCATED_VARIATION_ID = 0;
exports.UNALLOCATED_VARIATION_ID = UNALLOCATED_VARIATION_ID;
const OFF_VARIATION_KEY = 'off';
exports.OFF_VARIATION_KEY = OFF_VARIATION_KEY;
const KAMELEOON_COOKIE_KEY = 'kameleoonVisitorCode';
exports.KAMELEOON_COOKIE_KEY = KAMELEOON_COOKIE_KEY;
const VISITOR_CODE_LENGTH = 16;
exports.VISITOR_CODE_LENGTH = VISITOR_CODE_LENGTH;
const CACHE_CLEANUP_TIMEOUT = 10;
exports.CACHE_CLEANUP_TIMEOUT = CACHE_CLEANUP_TIMEOUT;
const CACHE_ITEM_LIFETIME = 5;
exports.CACHE_ITEM_LIFETIME = CACHE_ITEM_LIFETIME;
let Cookie = /*#__PURE__*/function (Cookie) {
  Cookie["PairsDelimiter"] = "; ";
  Cookie["KeyValueDelimiter"] = "=";
  Cookie["MaxAge"] = "Max-Age";
  Cookie["Path"] = "Path";
  Cookie["Domain"] = "Domain";
  return Cookie;
}({});
/**
 * @readonly
 * @enum {string} an Enum of all the possible Kameleoon SDK types
 * */
exports.Cookie = Cookie;
let SDKType = /*#__PURE__*/function (SDKType) {
  SDKType["NodeJS"] = "nodejs";
  SDKType["JavaScript"] = "javascript";
  SDKType["React"] = "react";
  return SDKType;
}({});
exports.SDKType = SDKType;
