// @kameleoon/javascript-sdk-core/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Browser", {
  enumerable: true,
  get: function () {
    return _kameleoonData.Browser;
  }
});
Object.defineProperty(exports, "BrowserType", {
  enumerable: true,
  get: function () {
    return _kameleoonData.BrowserType;
  }
});
Object.defineProperty(exports, "CLEANUP_TARGETING_DATA_INTERVAL", {
  enumerable: true,
  get: function () {
    return _campaignConfiguration.CLEANUP_TARGETING_DATA_INTERVAL;
  }
});
Object.defineProperty(exports, "ClientGetVisitorCodeParametersType", {
  enumerable: true,
  get: function () {
    return _types.ClientGetVisitorCodeParametersType;
  }
});
Object.defineProperty(exports, "Conversion", {
  enumerable: true,
  get: function () {
    return _kameleoonData.Conversion;
  }
});
Object.defineProperty(exports, "ConversionParametersType", {
  enumerable: true,
  get: function () {
    return _kameleoonData.ConversionParametersType;
  }
});
Object.defineProperty(exports, "Cookie", {
  enumerable: true,
  get: function () {
    return _constants.Cookie;
  }
});
Object.defineProperty(exports, "CustomData", {
  enumerable: true,
  get: function () {
    return _kameleoonData.CustomData;
  }
});
Object.defineProperty(exports, "Device", {
  enumerable: true,
  get: function () {
    return _kameleoonData.Device;
  }
});
Object.defineProperty(exports, "DeviceType", {
  enumerable: true,
  get: function () {
    return _kameleoonData.DeviceType;
  }
});
Object.defineProperty(exports, "Environment", {
  enumerable: true,
  get: function () {
    return _types.Environment;
  }
});
Object.defineProperty(exports, "Err", {
  enumerable: true,
  get: function () {
    return _tsRes.Err;
  }
});
Object.defineProperty(exports, "EventSourceOpenParametersType", {
  enumerable: true,
  get: function () {
    return _eventSource.EventSourceOpenParametersType;
  }
});
Object.defineProperty(exports, "ExperimentType", {
  enumerable: true,
  get: function () {
    return _types.ExperimentType;
  }
});
Object.defineProperty(exports, "ExternalPackageInfoType", {
  enumerable: true,
  get: function () {
    return _types.ExternalPackageInfoType;
  }
});
Object.defineProperty(exports, "FeatureFlagType", {
  enumerable: true,
  get: function () {
    return _types.FeatureFlagType;
  }
});
Object.defineProperty(exports, "FeatureVariableResultType", {
  enumerable: true,
  get: function () {
    return _types.FeatureVariableResultType;
  }
});
Object.defineProperty(exports, "GetClientConfigurationResultType", {
  enumerable: true,
  get: function () {
    return _requester.GetClientConfigurationResultType;
  }
});
Object.defineProperty(exports, "GetFeatureFlagVariableParamsType", {
  enumerable: true,
  get: function () {
    return _types.GetFeatureFlagVariableParamsType;
  }
});
Object.defineProperty(exports, "Header", {
  enumerable: true,
  get: function () {
    return _requester.Header;
  }
});
Object.defineProperty(exports, "HttpMethod", {
  enumerable: true,
  get: function () {
    return _requester.HttpMethod;
  }
});
Object.defineProperty(exports, "IExternalEventSource", {
  enumerable: true,
  get: function () {
    return _eventSource.IExternalEventSource;
  }
});
Object.defineProperty(exports, "IExternalEventSourceConstructor", {
  enumerable: true,
  get: function () {
    return _eventSource.IExternalEventSourceConstructor;
  }
});
Object.defineProperty(exports, "IExternalRequestDispatcher", {
  enumerable: true,
  get: function () {
    return _requester.IExternalRequestDispatcher;
  }
});
Object.defineProperty(exports, "IExternalStorage", {
  enumerable: true,
  get: function () {
    return _storage.IExternalStorage;
  }
});
Object.defineProperty(exports, "IExternalStorageConstructor", {
  enumerable: true,
  get: function () {
    return _storage.IExternalStorageConstructor;
  }
});
Object.defineProperty(exports, "IKameleoonCore", {
  enumerable: true,
  get: function () {
    return _kameleoonClient.IKameleoonClient;
  }
});
Object.defineProperty(exports, "InternalSDKConfigurationType", {
  enumerable: true,
  get: function () {
    return _types.InternalSDKConfigurationType;
  }
});
Object.defineProperty(exports, "JSONType", {
  enumerable: true,
  get: function () {
    return _campaignConfiguration.JSONType;
  }
});
Object.defineProperty(exports, "KameleoonCore", {
  enumerable: true,
  get: function () {
    return _kameleoonClient.KameleoonClient;
  }
});
Object.defineProperty(exports, "KameleoonCoreUtils", {
  enumerable: true,
  get: function () {
    return _kameleoonUtils.KameleoonUtils;
  }
});
Object.defineProperty(exports, "KameleoonDataType", {
  enumerable: true,
  get: function () {
    return _kameleoonData.KameleoonDataType;
  }
});
Object.defineProperty(exports, "KameleoonError", {
  enumerable: true,
  get: function () {
    return _kameleoonError.KameleoonError;
  }
});
Object.defineProperty(exports, "KameleoonException", {
  enumerable: true,
  get: function () {
    return _kameleoonError.KameleoonException;
  }
});
Object.defineProperty(exports, "KameleoonStorageKey", {
  enumerable: true,
  get: function () {
    return _storage.KameleoonStorageKey;
  }
});
Object.defineProperty(exports, "MobileGetVisitorCodeParametersType", {
  enumerable: true,
  get: function () {
    return _types.MobileGetVisitorCodeParametersType;
  }
});
Object.defineProperty(exports, "Ok", {
  enumerable: true,
  get: function () {
    return _tsRes.Ok;
  }
});
Object.defineProperty(exports, "PageView", {
  enumerable: true,
  get: function () {
    return _kameleoonData.PageView;
  }
});
Object.defineProperty(exports, "PageViewParametersType", {
  enumerable: true,
  get: function () {
    return _kameleoonData.PageViewParametersType;
  }
});
Object.defineProperty(exports, "RealTimeEventType", {
  enumerable: true,
  get: function () {
    return _eventSource.RealTimeEventType;
  }
});
Object.defineProperty(exports, "Result", {
  enumerable: true,
  get: function () {
    return _tsRes.Result;
  }
});
Object.defineProperty(exports, "SDKConfigurationType", {
  enumerable: true,
  get: function () {
    return _types.SDKConfigurationType;
  }
});
Object.defineProperty(exports, "SDKCoreParameters", {
  enumerable: true,
  get: function () {
    return _types.SDKCoreParameters;
  }
});
Object.defineProperty(exports, "SDKType", {
  enumerable: true,
  get: function () {
    return _constants.SDKType;
  }
});
Object.defineProperty(exports, "ServerGetVisitorCodeParametersType", {
  enumerable: true,
  get: function () {
    return _types.ServerGetVisitorCodeParametersType;
  }
});
Object.defineProperty(exports, "TrackConversionParamsType", {
  enumerable: true,
  get: function () {
    return _types.TrackConversionParamsType;
  }
});
Object.defineProperty(exports, "TrackParametersType", {
  enumerable: true,
  get: function () {
    return _requester.TrackParametersType;
  }
});
Object.defineProperty(exports, "VISITOR_CODE_LENGTH", {
  enumerable: true,
  get: function () {
    return _constants.VISITOR_CODE_LENGTH;
  }
});
Object.defineProperty(exports, "VariableType", {
  enumerable: true,
  get: function () {
    return _types.VariableType;
  }
});
var _kameleoonClient = require("./kameleoonClient");
var _kameleoonUtils = require("./kameleoonUtils");
var _kameleoonError = require("./kameleoonError");
var _storage = require("./storage");
var _constants = require("./constants");
var _campaignConfiguration = require("./campaignConfiguration");
var _tsRes = require("ts-res");
var _types = require("./types");
var _kameleoonData = require("./kameleoonData");
var _eventSource = require("./eventSource");
var _requester = require("./requester");
