// @kameleoon/javascript-sdk-core/dist/targeting/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "CustomDataCondition", {
  enumerable: true,
  get: function () {
    return _conditions.CustomDataCondition;
  }
});
Object.defineProperty(exports, "EvaluationDataType", {
  enumerable: true,
  get: function () {
    return _types.EvaluationDataType;
  }
});
Object.defineProperty(exports, "ExclusiveExperimentCondition", {
  enumerable: true,
  get: function () {
    return _conditions.ExclusiveExperimentCondition;
  }
});
Object.defineProperty(exports, "MatchTypeCustomData", {
  enumerable: true,
  get: function () {
    return _types.MatchTypeCustomData;
  }
});
Object.defineProperty(exports, "MatchTypeVariation", {
  enumerable: true,
  get: function () {
    return _types.MatchTypeVariation;
  }
});
Object.defineProperty(exports, "Operator", {
  enumerable: true,
  get: function () {
    return _types.Operator;
  }
});
Object.defineProperty(exports, "SegmentType", {
  enumerable: true,
  get: function () {
    return _types.SegmentType;
  }
});
Object.defineProperty(exports, "TargetExperimentCondition", {
  enumerable: true,
  get: function () {
    return _conditions.TargetExperimentCondition;
  }
});
Object.defineProperty(exports, "TargetingDataItemType", {
  enumerable: true,
  get: function () {
    return _types.TargetingDataItemType;
  }
});
Object.defineProperty(exports, "TargetingDataType", {
  enumerable: true,
  get: function () {
    return _types.TargetingDataType;
  }
});
Object.defineProperty(exports, "TargetingTree", {
  enumerable: true,
  get: function () {
    return _targetingTree.TargetingTree;
  }
});
Object.defineProperty(exports, "TargetingType", {
  enumerable: true,
  get: function () {
    return _types.TargetingType;
  }
});
var _targetingTree = require("./targetingTree");
var _types = require("./types");
var _conditions = require("./conditions");
