// @kameleoon/javascript-sdk-core/dist/targeting/types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TargetingType = exports.Operator = exports.MatchTypeVariation = exports.MatchTypeTest = exports.MatchTypeCustomData = void 0;
let Operator = /*#__PURE__*/function (Operator) {
  Operator["And"] = "AND";
  Operator["Or"] = "OR";
  return Operator;
}({});
exports.Operator = Operator;
let MatchTypeCustomData = /*#__PURE__*/function (MatchTypeCustomData) {
  MatchTypeCustomData["TRUE"] = "TRUE";
  MatchTypeCustomData["FALSE"] = "FALSE";
  MatchTypeCustomData["EXACT"] = "EXACT";
  MatchTypeCustomData["CONTAINS"] = "CONTAINS";
  MatchTypeCustomData["REGULAR_EXPRESSION"] = "REGULAR_EXPRESSION";
  MatchTypeCustomData["EQUAL"] = "EQUAL";
  MatchTypeCustomData["LOWER"] = "LOWER";
  MatchTypeCustomData["GREATER"] = "GREATER";
  MatchTypeCustomData["UNDEFINED"] = "UNDEFINED";
  MatchTypeCustomData["AMONG_VALUES"] = "AMONG_VALUES";
  return MatchTypeCustomData;
}({});
exports.MatchTypeCustomData = MatchTypeCustomData;
let MatchTypeVariation = /*#__PURE__*/function (MatchTypeVariation) {
  MatchTypeVariation["ANY"] = "ANY";
  MatchTypeVariation["EXACT"] = "EXACT";
  return MatchTypeVariation;
}({});
exports.MatchTypeVariation = MatchTypeVariation;
let MatchTypeTest = /*#__PURE__*/function (MatchTypeTest) {
  MatchTypeTest["ALL"] = "ALL";
  MatchTypeTest["TEST"] = "TEST";
  MatchTypeTest["PERSO"] = "PERSO";
  return MatchTypeTest;
}({});
exports.MatchTypeTest = MatchTypeTest;
let TargetingType = /*#__PURE__*/function (TargetingType) {
  TargetingType["PAGE_URL"] = "PAGE_URL";
  TargetingType["PAGE_TITLE"] = "PAGE_TITLE";
  TargetingType["LANDING_PAGE"] = "LANDING_PAGE";
  TargetingType["ORIGIN"] = "ORIGIN";
  TargetingType["ORIGIN_TYPE"] = "ORIGIN_TYPE";
  TargetingType["REFERRERS"] = "REFERRERS";
  TargetingType["NEW_VISITORS"] = "NEW_VISITORS";
  TargetingType["INTERESTS"] = "INTERESTS";
  TargetingType["BROWSER_LANGUAGE"] = "BROWSER_LANGUAGE";
  TargetingType["GEOLOCATION"] = "GEOLOCATION";
  TargetingType["DEVICE_TYPE"] = "DEVICE_TYPE";
  TargetingType["SCREEN_DIMENSION"] = "SCREEN_DIMENSION";
  TargetingType["VISITOR_IP"] = "VISITOR_IP";
  TargetingType["AD_BLOCKER"] = "AD_BLOCKER";
  TargetingType["PREVIOUS_PAGE"] = "PREVIOUS_PAGE";
  TargetingType["KEY_PAGES"] = "KEY_PAGES";
  TargetingType["PAGE_VIEWS"] = "PAGE_VIEWS";
  TargetingType["FIRST_VISIT"] = "FIRST_VISIT";
  TargetingType["LAST_VISIT"] = "LAST_VISIT";
  TargetingType["ACTIVE_SESSION"] = "ACTIVE_SESSION";
  TargetingType["TIME_SINCE_PAGE_LOAD"] = "TIME_SINCE_PAGE_LOAD";
  TargetingType["SAME_DAY_VISITS"] = "SAME_DAY_VISITS";
  TargetingType["VISITS"] = "VISITS";
  TargetingType["VISITS_BY_PAGE"] = "VISITS_BY_PAGE";
  TargetingType["INTERNAL_SEARCH_KEYWORDS"] = "INTERNAL_SEARCH_KEYWORDS";
  TargetingType["TABS_ON_SITE"] = "TABS_ON_SITE";
  TargetingType["CONVERSION_PROBABILITY"] = "CONVERSION_PROBABILITY";
  TargetingType["HEAT_SLICE"] = "HEAT_SLICE";
  TargetingType["SKY_STATUS"] = "SKY_STATUS";
  TargetingType["TEMPERATURE"] = "TEMPERATURE";
  TargetingType["DAY_NIGHT"] = "DAY_NIGHT";
  TargetingType["FORECAST_SKY_STATUS"] = "FORECAST_SKY_STATUS";
  TargetingType["FORECAST_TEMPERATURE"] = "FORECAST_TEMPERATURE";
  TargetingType["DAY_OF_WEEK"] = "DAY_OF_WEEK";
  TargetingType["TIME_RANGE"] = "TIME_RANGE";
  TargetingType["HOUR_MINUTE_RANGE"] = "HOUR_MINUTE_RANGE";
  TargetingType["JS_CODE"] = "JS_CODE";
  TargetingType["COOKIE"] = "COOKIE";
  TargetingType["EVENT"] = "EVENT";
  TargetingType["BROWSER"] = "BROWSER";
  TargetingType["OPERATING_SYSTEM"] = "OPERATING_SYSTEM";
  TargetingType["DOM_ELEMENT"] = "DOM_ELEMENT";
  TargetingType["MOUSE_OUT"] = "MOUSE_OUT";
  TargetingType["TARGET_EXPERIMENT"] = "TARGET_EXPERIMENT";
  TargetingType["TARGET_PERSONALIZATION"] = "TARGET_PERSONALIZATION";
  TargetingType["EXCLUSIVE_EXPERIMENT"] = "EXCLUSIVE_EXPERIMENT";
  TargetingType["CONVERSIONS"] = "CONVERSIONS";
  TargetingType["CUSTOM_DATUM"] = "CUSTOM_DATUM";
  TargetingType["YSANCE_SEGMENT"] = "YSANCE_SEGMENT";
  TargetingType["YSANCE_ATTRIBUT"] = "YSANCE_ATTRIBUT";
  TargetingType["TEALIUM_BADGE"] = "TEALIUM_BADGE";
  TargetingType["TEALIUM_AUDIENCE"] = "TEALIUM_AUDIENCE";
  TargetingType["PRICE_OF_DISPLAYED_PAGE"] = "PRICE_OF_DISPLAYED_PAGE";
  TargetingType["NUMBER_OF_VISITED_PAGES"] = "NUMBER_OF_VISITED_PAGES";
  TargetingType["VISITED_PAGES"] = "VISITED_PAGES";
  TargetingType["CUSTOM"] = "CUSTOM";
  TargetingType["EXPLICIT_TRIGGER"] = "EXPLICIT_TRIGGER";
  return TargetingType;
}({});
exports.TargetingType = TargetingType;
