// @kameleoon/javascript-sdk-core/dist/campaignConfiguration/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "CLEANUP_TARGETING_DATA_INTERVAL", {
  enumerable: true,
  get: function () {
    return _constants.CLEANUP_TARGETING_DATA_INTERVAL;
  }
});
Object.defineProperty(exports, "CampaignConfiguration", {
  enumerable: true,
  get: function () {
    return _campaignConfiguration.CampaignConfiguration;
  }
});
Object.defineProperty(exports, "CampaignConfigurationType", {
  enumerable: true,
  get: function () {
    return _types.CampaignConfigurationType;
  }
});
Object.defineProperty(exports, "ConfigurationDataType", {
  enumerable: true,
  get: function () {
    return _types.ConfigurationDataType;
  }
});
Object.defineProperty(exports, "ExperimentItemType", {
  enumerable: true,
  get: function () {
    return _types.ExperimentItemType;
  }
});
Object.defineProperty(exports, "FeatureFlagType", {
  enumerable: true,
  get: function () {
    return _types.FeatureFlagType;
  }
});
Object.defineProperty(exports, "FeatureVariableType", {
  enumerable: true,
  get: function () {
    return _types.FeatureVariableType;
  }
});
Object.defineProperty(exports, "JSONType", {
  enumerable: true,
  get: function () {
    return _types.JSONType;
  }
});
Object.defineProperty(exports, "JSONValue", {
  enumerable: true,
  get: function () {
    return _types.JSONValue;
  }
});
Object.defineProperty(exports, "RespoolTimeType", {
  enumerable: true,
  get: function () {
    return _types.RespoolTimeType;
  }
});
Object.defineProperty(exports, "RuleItemType", {
  enumerable: true,
  get: function () {
    return _types.RuleItemType;
  }
});
Object.defineProperty(exports, "RuleType", {
  enumerable: true,
  get: function () {
    return _types.RuleType;
  }
});
Object.defineProperty(exports, "VariationByExpositionType", {
  enumerable: true,
  get: function () {
    return _types.VariationByExpositionType;
  }
});
var _types = require("./types");
var _campaignConfiguration = require("./campaignConfiguration");
var _constants = require("./constants");
