// @kameleoon/javascript-sdk-core/dist/campaignConfiguration/types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RuleType = exports.FeatureStatus = exports.FeatureFlagSdkLanguageType = exports.ExperimentType = exports.ExperimentStatus = void 0;
let ExperimentType = /*#__PURE__*/function (ExperimentType) {
  ExperimentType["CLASSIC"] = "CLASSIC";
  ExperimentType["SERVER_SIDE"] = "SERVER_SIDE";
  ExperimentType["DEVELOPER"] = "DEVELOPER";
  ExperimentType["MVT"] = "MVT";
  ExperimentType["HYBRID"] = "HYBRID";
  return ExperimentType;
}({});
exports.ExperimentType = ExperimentType;
let RuleType = /*#__PURE__*/function (RuleType) {
  RuleType["PROGRESSIVE_DELIVERY"] = "PROGRESSIVE_DELIVERY";
  RuleType["TARGETED_DELIVERY"] = "TARGETED_DELIVERY";
  RuleType["EXPERIMENTATION"] = "EXPERIMENTATION";
  return RuleType;
}({});
exports.RuleType = RuleType;
let FeatureFlagSdkLanguageType = /*#__PURE__*/function (FeatureFlagSdkLanguageType) {
  FeatureFlagSdkLanguageType["ANDROID"] = "ANDROID";
  FeatureFlagSdkLanguageType["SWIFT"] = "SWIFT";
  FeatureFlagSdkLanguageType["JAVA"] = "JAVA";
  FeatureFlagSdkLanguageType["CSHARP"] = "CSHARP";
  FeatureFlagSdkLanguageType["NODEJS"] = "NODEJS";
  FeatureFlagSdkLanguageType["PHP"] = "PHP";
  FeatureFlagSdkLanguageType["RUBY"] = "RUBY";
  FeatureFlagSdkLanguageType["GO"] = "GO";
  FeatureFlagSdkLanguageType["FLUTTER"] = "FLUTTER";
  FeatureFlagSdkLanguageType["REACTJS"] = "REACTJS";
  return FeatureFlagSdkLanguageType;
}({});
exports.FeatureFlagSdkLanguageType = FeatureFlagSdkLanguageType;
let FeatureStatus = /*#__PURE__*/function (FeatureStatus) {
  FeatureStatus["ACTIVATED"] = "ACTIVATED";
  FeatureStatus["DEACTIVATED"] = "DEACTIVATED";
  FeatureStatus["SCHEDULED"] = "SCHEDULED";
  return FeatureStatus;
}({});
exports.FeatureStatus = FeatureStatus;
let ExperimentStatus = /*#__PURE__*/function (ExperimentStatus) {
  ExperimentStatus["DRAFT"] = "DRAFT";
  ExperimentStatus["ACTIVE"] = "ACTIVE";
  ExperimentStatus["PAUSED"] = "PAUSED";
  ExperimentStatus["SCHEDULED"] = "SCHEDULED";
  ExperimentStatus["STOPPED"] = "STOPPED";
  ExperimentStatus["USED_AS_PERSONALIZATION"] = "USED_AS_PERSONALIZATION";
  ExperimentStatus["DEVIATED"] = "DEVIATED";
  return ExperimentStatus;
}({});
exports.ExperimentStatus = ExperimentStatus;
