// @kameleoon/javascript-sdk-core/dist/kameleoonData/browser.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Browser = void 0;
var _nonce = require("./nonce");
var _requester = require("../requester");
var _types = require("./types");
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return typeof key === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (typeof input !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (typeof res !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
/**
 * @class
 * Browser - a class for creating an instance for browser of a certain type
 * */
class Browser extends _nonce.Nonce {
  /**
   * @param {BrowserType} browser - one of the browser types available for data, if browser can not be found in a list, it can be marked as `BrowserType.Other`
   * */
  constructor(browser) {
    super();
    _defineProperty(this, "browser", void 0);
    this.browser = browser;
  }

  /**
   * @property url - url for the current data type used for events tracking
   * */
  get url() {
    return _requester.UrlEventType.StaticData + _requester.UrlParameter.BrowserIndex + this.browser + _requester.UrlParameter.Nonce + this.nonce;
  }

  /**
   * @property data - data of the current class instance
   * */
  get data() {
    return {
      browser: this.browser,
      type: _types.KameleoonData.Browser,
      nonce: this.nonce
    };
  }
}
exports.Browser = Browser;
