// @kameleoon/javascript-sdk-core/dist/kameleoonData/conversion.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Conversion = void 0;
var _requester = require("../requester");
var _nonce = require("./nonce");
var _types = require("./types");
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return typeof key === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (typeof input !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (typeof res !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
/**
 * @class
 * Conversion - a class for creating an instance for conversion tracking
 * */
class Conversion extends _nonce.Nonce {
  /**
   * @param {ConversionParametersType} converstionParameters - an object with conversion parameters
   * */
  constructor({
    goalId,
    revenue = 0,
    negative = false
  }) {
    super();
    _defineProperty(this, "goalId", void 0);
    _defineProperty(this, "revenue", void 0);
    _defineProperty(this, "negative", void 0);
    this.goalId = goalId;
    this.revenue = revenue;
    this.negative = negative;
  }

  /**
   * @property url - url for the current data type used for events tracking
   * */
  get url() {
    return _requester.UrlEventType.Conversion + _requester.UrlParameter.GoalId + String(this.goalId) + _requester.UrlParameter.Revenue + String(this.revenue) + _requester.UrlParameter.Negative + String(this.negative) + _requester.UrlParameter.Nonce + this.nonce;
  }

  /**
   * @property data - data of the current class instance
   * */
  get data() {
    return {
      goalId: this.goalId,
      revenue: this.revenue,
      negative: this.negative,
      type: _types.KameleoonData.Conversion,
      nonce: this.nonce
    };
  }
}
exports.Conversion = Conversion;
