// @kameleoon/javascript-sdk-core/dist/eventSource/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "EventSourceOpenParametersType", {
  enumerable: true,
  get: function () {
    return _types.EventSourceOpenParametersType;
  }
});
Object.defineProperty(exports, "ExternalEventSource", {
  enumerable: true,
  get: function () {
    return _externalEventSource.ExternalEventSource;
  }
});
Object.defineProperty(exports, "IEventSource", {
  enumerable: true,
  get: function () {
    return _externalEventSource.IEventSource;
  }
});
Object.defineProperty(exports, "IExternalEventSource", {
  enumerable: true,
  get: function () {
    return _types.IExternalEventSource;
  }
});
Object.defineProperty(exports, "IExternalEventSourceConstructor", {
  enumerable: true,
  get: function () {
    return _types.IExternalEventSourceConstructor;
  }
});
Object.defineProperty(exports, "REAL_TIME_EVENT", {
  enumerable: true,
  get: function () {
    return _constants.REAL_TIME_EVENT;
  }
});
Object.defineProperty(exports, "RealTimeEventType", {
  enumerable: true,
  get: function () {
    return _types.RealTimeEventType;
  }
});
var _types = require("./types");
var _externalEventSource = require("./externalEventSource");
var _constants = require("./constants");
