// @kameleoon/javascript-sdk-core/dist/requester/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UrlTracking = exports.UrlQuery = exports.UrlParameter = exports.UrlEventType = exports.URL = exports.Header = void 0;
let Header = /*#__PURE__*/function (Header) {
  Header["UserAgent"] = "User-Agent";
  return Header;
}({});
exports.Header = Header;
const topLevelDomain = undefined === 'test' ? 'net' : 'com';
const dataTopLevelDomain = undefined === 'test' ? 'net' : 'io';
const UrlEventType = {
  CustomData: 'eventType=customData',
  StaticData: 'eventType=staticData',
  Page: 'eventType=page',
  Conversion: 'eventType=conversion',
  Activity: 'eventType=activity',
  Experiment: 'eventType=experiment'
};
exports.UrlEventType = UrlEventType;
const UrlParameter = {
  Title: '&title=',
  ReferrersIndices: '&referrersIndices=',
  Negative: '&negative=',
  Revenue: '&revenue=',
  Overwrite: '&overwrite=',
  Index: '&index=',
  BrowserIndex: '&browserIndex=',
  Href: '&href=',
  DeviceType: '&deviceType=',
  SiteCode: '&siteCode=',
  GoalId: '&goalId=',
  VisitorCode: '&visitorCode=',
  Environment: '&environment=',
  VariationId: '&variationId=',
  Ts: '&ts=',
  Key: '&key=',
  SdkName: '&sdkName=',
  SdkVersion: '&sdkVersion=',
  ValuesCountMap: '&valuesCountMap=',
  Nonce: '&nonce=',
  Id: '&id='
};
exports.UrlParameter = UrlParameter;
const UrlQuery = {
  Sse: 'sse?siteCode=',
  Mobile: 'mobile?siteCode=',
  Map: 'map?siteCode=',
  Events: 'events?siteCode='
};
exports.UrlQuery = UrlQuery;
const UrlTracking = {
  Visit: 'visit/',
  Map: 'map/'
};
exports.UrlTracking = UrlTracking;
const URL = {
  EVENTS: `https://events.kameleoon.${topLevelDomain}:8110/`,
  TRACKING: `https://data.kameleoon.${dataTopLevelDomain}/${UrlTracking.Visit + UrlQuery.Events}`,
  DATA: `https://data.kameleoon.${dataTopLevelDomain}/${UrlTracking.Map + UrlQuery.Map}`,
  CLIENT_CONFIGURATION: `https://client-config.kameleoon.${topLevelDomain}/`
};
exports.URL = URL;
